### PR TITLE
【translate】doc/en/client_en.md

### PR DESCRIPTION
**Pull request by SeanHai**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?): 

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: docs/en/client_en.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+1. File chunk -> chunk id: The mapping between logical chunks user's file and the physical chunks on chunk server.
+2. Chunk id -> copyset id(raft group): The Raft group that those physical chunks belong to.
+3. Copyset id -> serverlist: Chunk server list that Raft groups belong to.
+
+### 2.5 MDS Failover Supports
+
+MDS provides metadata storage services, and serves in cluster for improving availability. Only one MDS provides services at the same time, and others will monitor through Etcd, getting ready to replace the leader at any time.
+
+Curve client will communicate with MDS by RPC when I/O is dispatched and the control panel is called, during which the  address of corresponding MDS is required. Therefore, the clients also needs to change the address if the service node in MDS cluster is switched.
```
Comment (wu-hanqing) [inline docs/en/client_en.md:56]: by RPC -> through RPC
control panel -> control request
corresponding MDS -> working MDS

File: docs/en/client_en.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+</p>
+
+### 3.2 Thread Module
+
+<p align="center">
+    <img src="../images/curve-client-thread-model.png" alt="curve-client-thread-model.png" width="600" /><br>
+    <font size=3>Figure 3: Thread module of Curve client</font>
+</p>
+
+Figure 3 shows the thread module of Curve client. Use asynchronous I/O request as an example, for the client there are 4 kinds of threads involved:
```
Comment (wu-hanqing) [inline docs/en/client_en.md:86]: thread module -> thread model

File: docs/en/client_en.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+1. User thread: When a user initiates an asynchronous I/O request, AioWrite interface of Curve client will be called. AioWrite will encapsulate user's request and place it into a task queue, and this user call will return at this stage.
```
Comment (wu-hanqing) [inline docs/en/client_en.md:88]: place it into a task queue -> push it

File: docs/en/client_en.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+2. Task queue thread: This thread will monitor the task queue, and once an I/O task enter, it will fetch from the queue and split the I/O into sub-I/Os and place them into the I/O dispatching queue.
```
Comment (wu-hanqing) [inline docs/en/client_en.md:89]: place -> push

File: docs/en/client_en.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+Curve client trigger metadata update also by RPC. The metadata being updated including:
+
+1. Leader information of the Raft group
+
+   The request of the client will be sent to the leader of the Raft group, and the client will fetch the leader info of current Raft group (cached in meta cache). If a request was sent to the chunk server that is no longer the leader, the client will initiate a GetLeader request to other chunk servers in current Raft group. This request can be sent to any of them since every node will know the leader if a new one has been elected. After knowing the new leader, the client will update the meta cache, then resend the request to the new leader.
+
+2. Peer info in Raft group
+
+   In such an extreme case that all peers in a Raft group are change due to unexpected issues like node failure, GetLeader request will not able to get the real leader. Client will fetch the new Raft group info from MDS in this case after retried for many times.
```
Comment (wu-hanqing) [inline docs/en/client_en.md:116]: many times -> some times

File: docs/en/client_en.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
+[中文版](../cn/curve-client.md)
+
+# Curve Client
+
+## 1 Outline
+
+As a centralized distributed storage system, Curve consists of Client module for user's requests, Chunkserver module for basic storage, metadata server MDS and snapshot system.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:7]: > Curve consist of Client which is used for processing user's requests, Chunkserver for underlying data storage

done

File: docs/en/client_en.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+Curve client serves the upper layer by dynamic library libcurve, which can be linked by external processes like QEMU and Curve-NBD for utilizing services provided by Curve (Curve client will no longer provide services directly to QEMU and Curve-NBD after the hot upgrade, see [NEBD](../en/nebd_en.md) for more details). Thus, Curve client can be regarded as the entrance of user's operations.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:9]: > As a library, Curve Client can be linked by other processes

to : As a library, Curve Client can be linked by other processes like QEMU and Curve-NBD for utilizing services provided by Curve (Curve client will no longer provide services directly to QEMU and Curve-NBD after the hot upgrade, see [NEBD](../en/nebd_en.md) for more details). Thus, Curve client can be regarded as the entrance of user's operations.

File: docs/en/client_en.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+## 2 Curve Client Functions
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:11]: > Curve Client Function Points

done

File: docs/en/client_en.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+
+Curve provides block storage service based on its underlying distributed file storage system, and Curve client provides user data storage services interfaces. From user's aspect, what Curve have provided is a block device capable for reading and writing randomly, which corresponds to a file in CurveFS that stored on different nodes of the cluster. To make it clear, let's use an example to illustrate this idea:
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:13]: > based on its underlying distributed filesystem, and Curve client provides block device interfaces. From the user's perspective ...
> 
> 最后的idea可以去掉

done

File: docs/en/client_en.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+When a user create a block device of 1T, this block device corresponds to a 1T size file on CurveFS.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:15]: > When the user creates a 1T block device ...

done

File: docs/en/client_en.md [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+<font size=3>Figure 1: Example of the mapping between user's block device and its storage</font>
+</p>
+
+For the user, the 1T block device is a continuous address for reading and writing. But from Curve's side, this block device is a space consists of many chunks with unique ID, and every chunk corresponds to a piece of physical space on Chunkserver. Figure 1 above illustrates this relationship.
+
+Curve client communicate with storage cluster by RPC, and from the description above we can try to make a conclusion on what Curve client should do:
+
+### 2.1 Providing Interface
+
+Users operate block devices using interface provided by Curve client, including creating, opening, reading and writing. For achieving the best performance, asynchronous reading/writing interfaces are provided.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:30]: > User operates ...
> 
> including creating, opening, reading and writing -> including create, open, read and write.
> 
> asynchronous reading/writing -> asynchronous read/write

done

File: docs/en/client_en.md [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+### 2.2 I/O Splitting
+
+From figure 1 we can see, user's block device is actually split and stored in nodes by chunks, this is for higher concurrency and data reliability.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:34]: > As we can see from Figure 1, user's block devices are scattered on multiple storage nodes in the underlying physical storage in chunks,

done

File: docs/en/client_en.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+When writing to Curve client, the client will receive the offset, length and actual data of the I/O. The corresponding space of offset + length may cover more than one chunk, and chunks may be stored in different chunk server, so the Curve client should transfer user's block device I/O request to the chunk request toward different chunk server and dispatch.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:36]: > chunks may be stored in different chunk server -> chunks may be placed in different Chunkserver

done

File: docs/en/client_en.md [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+
+### 2.3 I/O Tracking
+
+I/O are split according to the offset and length. For example, if a user send an I/O request with offset = 0 and length = 2 * chunk size, this request will be split into at least two when arrive the client since the I/O target space crosses two chunks on different chunk server. In the client, both the I/O interface and writing chunk are asynchronous, and as a result tracking every split I/Os and record their returns are necessary. The request will return to the user only when every split I/Os have returned.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:40]: > this request will be split into at least two when arrive the client -> this request will be split into at least two when it arrives at the client
> 
> `chunk server`统一改成`Chunkserver`吧。
> 
> split I/Os -> splitted I/Os

done

File: docs/en/client_en.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+
+The metadata that a user I/O requires including:
+
+1. File chunk -> chunk id: The mapping between logical chunks user's file and the physical chunks on chunk server.
+2. Chunk id -> copyset id(raft group): The Raft group that those physical chunks belong to.
+3. Copyset id -> serverlist: Chunk server list that Raft groups belong to.
+
+### 2.5 MDS Failover Supports
+
+MDS provides metadata storage services, and serves in cluster for improving availability. Only one MDS provides services at the same time, and others will monitor through Etcd, getting ready to replace the leader at any time.
```
[the remark this answers is not in the document]
Comment (SeanHai) [reply docs/en/client_en.md:54]: > MDS provides metadata services, and in order to improve availability, mds serves as a cluster.

done

File: docs/en/client_en.md (hunk 1 above)
Comment (SeanHai) [reply docs/en/client_en.md:56]: > by RPC -> through RPC
> control panel -> control request
> corresponding MDS -> working MDS

done

File: docs/en/client_en.md (hunk 2 above)
Comment (SeanHai) [reply docs/en/client_en.md:86]: > thread module -> thread model

done

File: docs/en/client_en.md (hunk 3 above)
Comment (SeanHai) [reply docs/en/client_en.md:88]: > place it into a task queue -> push it

done

File: docs/en/client_en.md (hunk 4 above)
Comment (SeanHai) [reply docs/en/client_en.md:89]: > place -> push

done

File: docs/en/client_en.md (hunk 5 above)
Comment (SeanHai) [reply docs/en/client_en.md:116]: > many times -> some times

done

**Review by wu-hanqing — APPROVED** [review]

**Review by ilixiaocui — APPROVED** [review]

**Pull request merged**